napari/components/add_layers_mixin.py

```diff
@@ -88,7 +88,7 @@ def add_image(
 
         Parameters
         ----------
-        data : array or list of array
+        data : np.ndarray or list of np.ndarray
```

Comment:
@Czaki the problem is that this is inaccurate. We actually accept anything coercible to a NumPy array here, including dask arrays, pint Quantities, zarr arrays, pytorch CPU tensors...

Reply:
Hm. I see  [PEP 544](https://www.python.org/dev/peps/pep-0544/) which can be used to solve this. 

I found that the problem with is produced by existence of `array` module https://docs.python.org/3/library/array.html and name collision,

Reply:
yes indeed, pep 544 holds promise... and has been a hot topic of conversation within the napari team for a while now! :)

it is definitely a long term goal to implement real type stubs for valid napari layer data types.  (see @jni's [imtypes](https://github.com/napari/image-types) module for a _very_ early start on this concept... though that was created long before we had started discussing Protocols).  After digging in a little deeper, it turns out to be a bit harder than it first appears... and so we have a temporary "copout" union [here](https://github.com/napari/napari/blob/6933f5bcc12608d66458987d69db2f7c535b21de/napari/types.py#L13-L20).

but if you wanted to look into creating a protocol type that would recognize all of the various types of data that napari supports, such that an IDE recognizes them, it is definitely something that is needed!

Reply:
But If I good understand Protocols are introduced in python 3.8. And It takes long time before `napari` drop support for older pythons? 

(PyCharm suggest that support this from python 3.7).

Also connected with this topisc. I do not see type annotation in napari code. But @tlambert03 message suggest that there is no contradictions to introduce them?

Reply:
> But If I good understand Protocols are introduced in python 3.8. And It takes long time before `napari` drop support for older pythons?

there is a backport [typing-extensions](https://pypi.org/project/typing-extensions/) that covers all the versions we support.


> Also connected with this topisc. I do not see type annotation in napari code. But @tlambert03 message suggest that there is no contradictions to introduce them?

there are plenty of type annotations in napari code!  ([31 different files have type hints](https://github.com/napari/napari/search?q=%22from+typing+import%22)) though of course they are not everywhere, and in particular, they are not all over our layer classes (though, see https://github.com/napari/napari/pull/1097 for a start and some discussion on it)

by all means, feel free to use type annotations... but if you use one that is introduced after 3.6, we will need to depend on the backport (which we have not yet done)

Reply:
Note that `imtypes` was written before PEP-544 was accepted. It would probably look different now. =)

We will drop Py3.6 support in November at the latest, along with the rest of the scientific Python stack, see [NEP-29](https://numpy.org/neps/nep-0029-deprecation_policy.html). So we can definitely start to contemplate things that are available in Py3.7 only. But I'm 100% on board with depending on typing backports because I love typing. =)